README.rst
Facebook Auth support in Kinto
==============================

|travis| |master-coverage|

.. |travis| image:: https://travis-ci.org/Kinto/kinto-facebook.svg?branch=master
    :target: https://travis-ci.org/Kinto/kinto-facebook

.. |master-coverage| image::
    https://coveralls.io/repos/Kinto/kinto-facebook/badge.png?branch=master
    :alt: Coverage
    :target: https://coveralls.io/r/Kinto/kinto-facebook

*Kinto-facebook* enables authentication in *Kinto* applications using
*Facebook* OAuth2 bearer tokens.

It provides:

* An authentication policy class;
* Integration with *Kinto* cache backend for token verifications;
* Integration with *Kinto* for heartbeat view checks;
* Some optional endpoints to perform the *OAuth* dance (*optional*).


* `Kinto documentation <http://kinto.readthedocs.io/en/latest/>`_
* `Issue tracker <https://github.com/Kinto/kinto-facebook/issues>`_


Installation
------------

First of all, `create a Facebook Auth app <https://developers.facebook.com/apps/>`_.

You need to grab the `client_id`, `client_secret` and
`app_client_token` in the app general settings page:

::

   https://developers.facebook.com/apps/{client_id}/settings/basic/
   https://developers.facebook.com/apps/{client_id}/settings/advanced/


Install the Python package:

::

    pip install kinto-facebook


Include the package in the project configuration:

::

    kinto.includes = kinto_facebook

And configure authentication policy using `pyramid_multiauth
<https://github.com/mozilla-services/pyramid_multiauth#deployment-settings>`_ formalism:

::

    multiauth.policies = facebook
    multiauth.policy.facebook.use = kinto_facebook.authentication.FacebookAuthenticationPolicy

By default, it will rely on the cache configured in *Kinto*.


Configuration
-------------

Fill those settings with the values obtained during the application registration:

::

    facebook.client_id = 89513028159972bc
    facebook.client_secret = 9aced230585cc0aaea0a3467dd800
    facebook.app_access_token = 30ce836499a146428464b28816b1618b
    facebook.webapp.authorized_domains = *
    # facebook.cache_ttl_seconds = 300
    # facebook.state.ttl_seconds = 3600


In case the application shall not behave as a relier (a.k.a. OAuth dance
endpoints disabled):

::

    facebook.relier.enabled = false


If necessary, override default values for authentication policy:

::

    facebook.header_type = Bearer+Facebook


Login flow
----------

OAuth Bearer token
::::::::::::::::::

Use the OAuth token with this header:

::

    Authorization: Bearer <access_token>


:notes:

    If the token is not valid, this will result in a ``401`` error response.


Obtain token using Web UI
:::::::::::::::::::::::::

* Navigate the client to ``GET /facebook/login?redirect=http://app-endpoint/%23``.
  There, a session cookie will be set, and the client will be redirected to a login
  form on the Facebook content server;
* After submitting the credentials on the login page, the client will
  be redirected to ``http://app-endpoint/#{token}`` (the web-app).


CHANGELOG.rst
Changelog
=========

This document describes changes between each past release.

1.0.0 (unreleased)
------------------

- Nothing changed yet.


0.1.0 (2018-03-15)
------------------

- Initial implementation.


tests/test_views.py
import mock
import requests
from requests.exceptions import HTTPError
import unittest
import webtest
from urllib.parse import parse_qs, urlparse

import kinto.core
from kinto.core.errors import ERRORS
from kinto.core.testing import FormattedErrorMixin
from kinto.core.utils import random_bytes_hex
from pyramid.config import Configurator
from time import sleep

from kinto_facebook import __version__ as kinto_facebook_version


def get_request_class(prefix):

    class PrefixedRequestClass(webtest.app.TestRequest):

        @classmethod
        def blank(cls, path, *args, **kwargs):
            path = '/%s%s' % (prefix, path)
            return webtest.app.TestRequest.blank(path, *args, **kwargs)

    return PrefixedRequestClass


class BaseWebTest(object):
    """Base Web Test to test your cornice service.

    It setups the database before each test and delete it after.
    """

    api_prefix = "v1"

    def __init__(self, *args, **kwargs):
        super(BaseWebTest, self).__init__(*args, **kwargs)
        self.app = self._get_test_app()
        self.headers = {
            'Content-Type': 'application/json',
        }
        self._facebook_verify_patcher = mock.patch('kinto_facebook.authentication.requests')

    def setUp(self):
        super(BaseWebTest, self).setUp()
        self.facebook_verify = self._facebook_verify_patcher.start()
        self.facebook_verify.get.return_value.json.return_value = {
            'data': {
                'user_id': 'bob'
            }
        }

    def tearDown(self):
        super(BaseWebTest, self).tearDown()
        self._facebook_verify_patcher.stop()

    def _get_test_app(self, settings=None):
        config = self._get_app_config(settings)
        wsgi_app = config.make_wsgi_app()
        app = webtest.TestApp(wsgi_app)
        app.RequestClass = get_request_class(self.api_prefix)
        return app

    def _get_app_config(self, settings=None):
        config = Configurator(settings=self.get_app_settings(settings))
        kinto.core.initialize(config, version='1.0.0')
        return config

    def get_app_settings(self, additional_settings=None):
        settings = kinto.core.DEFAULT_SETTINGS.copy()
        settings['project_name'] = 'kinto'
        settings['includes'] = 'kinto_facebook'
        settings['multiauth.policies'] = 'facebook'
        authn = 'kinto_facebook.authentication.FacebookAuthenticationPolicy'
        settings['multiauth.policy.facebook.use'] = authn
        settings['cache_backend'] = 'kinto.core.cache.memory'
        settings['userid_hmac_secret'] = random_bytes_hex(16)
        settings['facebook.relier.enabled'] = True
        settings['facebook.webapp.authorized_domains'] = ['*.firefox.com']

        if additional_settings is not None:
            settings.update(additional_settings)
        return settings


class DeactivatedRelierTest(BaseWebTest, unittest.TestCase):
    def get_app_settings(self, additional_settings=None):
        settings = super(DeactivatedRelierTest, self).get_app_settings()
        settings['facebook.relier.enabled'] = False
        return settings

    def test_login_view_is_not_available(self):
        self.app.get('/facebook/login', status=404)

    def test_token_view_is_not_available(self):
        self.app.get('/facebook/token', status=404)


class LoginViewTest(BaseWebTest, unittest.TestCase):
    url = '/facebook/login?redirect=https://readinglist.firefox.com'

    def get_app_settings(self, additional_settings=None):
        additional_settings = additional_settings or {}
        additional_settings.update({
            'facebook.client_id': 'cid',
            'facebook.client_secret': 'csecret',
            'facebook.app_access_token': 'app_token'
        })
        return super(LoginViewTest, self).get_app_settings(additional_settings)

    def test_redirect_parameter_is_mandatory(self):
        url = '/facebook/login'
        r = self.app.get(url, status=400)
        self.assertIn('redirect', r.json['message'])

    def test_redirect_parameter_should_be_refused_if_not_whitelisted(self):
        url = '/facebook/login?redirect=http://not-whitelisted.tld'
        r = self.app.get(url, status=400)
        self.assertIn('redirect', r.json['message'])

    def test_redirect_parameter_should_be_accepted_if_whitelisted(self):
        with mock.patch.dict(self.app.app.registry.settings,
                             [('facebook.webapp.authorized_domains',
                               '*.whitelist.ed')]):
            url = '/facebook/login?redirect=http://iam.whitelist.ed'
            self.app.get(url)

    def test_redirect_parameter_should_be_rejected_if_no_whitelist(self):
        with mock.patch.dict(self.app.app.registry.settings,
                             [('facebook.webapp.authorized_domains',
                               '')]):
            url = '/facebook/login?redirect=http://iam.whitelist.ed'
            r = self.app.get(url, status=400)
        self.assertIn('redirect', r.json['message'])

    def test_login_view_persists_state(self):
        r = self.app.get(self.url)
        url = r.headers['Location']
        url_fragments = urlparse(url)
        queryparams = parse_qs(url_fragments.query)
        state = queryparams['state'][0]
        self.assertEqual(self.app.app.registry.cache.get(state),
                         'https://readinglist.firefox.com')

    def test_login_view_persists_state_with_expiration(self):
        r = self.app.get(self.url)
        url = r.headers['Location']
        url_fragments = urlparse(url)
        queryparams = parse_qs(url_fragments.query)
        state = queryparams['state'][0]
        self.assertGreater(self.app.app.registry.cache.ttl(state), 299)
        self.assertLessEqual(self.app.app.registry.cache.ttl(state), 300)

    def test_login_view_persists_state_with_expiration_from_settings(self):
        r = self.app.get(self.url)
        url = r.headers['Location']
        url_fragments = urlparse(url)
        queryparams = parse_qs(url_fragments.query)
        state = queryparams['state'][0]
        self.assertGreater(self.app.app.registry.cache.ttl(state), 299)
        self.assertLessEqual(self.app.app.registry.cache.ttl(state), 300)

    @mock.patch('kinto_facebook.views.uuid.uuid4')
    def test_login_view_redirects_to_authorization(self, mocked_uuid):
        mocked_uuid.return_value = mock.MagicMock(hex='1234')
        settings = self.app.app.registry.settings
        auth_endpoint = settings.get('facebook.authorization_endpoint')
        client_id = settings.get('facebook.client_id')
        redirect_uri = 'http%3A%2F%2Flocalhost%2Fv1%2Ffacebook%2Ftoken'
        expected_redirect = '{}?client_id={}&redirect_uri={}&state=1234'.format(auth_endpoint,
                                                                                client_id,
                                                                                redirect_uri)

        r = self.app.get(self.url)
        self.assertEqual(r.status_code, 302)
        self.assertEqual(r.headers['Location'], expected_redirect)


class TokenViewTest(FormattedErrorMixin, BaseWebTest, unittest.TestCase):
    url = '/facebook/token'
    login_url = '/facebook/login?redirect=https://readinglist.firefox.com'

    def __init__(self, *args, **kwargs):
        super(TokenViewTest, self).__init__(*args, **kwargs)
        self._facebook_trade_patcher = mock.patch('kinto_facebook.views.requests')

    def setUp(self):
        super(BaseWebTest, self).setUp()
        self.facebook_trade = self._facebook_trade_patcher.start()
        self.facebook_trade.get.return_value.json.return_value = {"access_token": "oauth-token"}

    def tearDown(self):
        super(BaseWebTest, self).tearDown()
        self._facebook_trade_patcher.stop()

    def test_fails_if_no_ongoing_session(self):
        url = '{url}?state=abc&code=1234'.format(url=self.url)
        resp = self.app.get(url, status=408)
        error_msg = 'The Facebook Auth session was not found, please re-authenticate.'
        self.assertFormattedError(
            resp, 408, ERRORS.MISSING_AUTH_TOKEN, "Request Timeout", error_msg)

    def test_fails_if_state_or_code_is_missing(self):
        headers = {'Cookie': 'state=abc'}
        for params in ['', '?state=abc', '?code=1234']:
            r = self.app.get(self.url + params, headers=headers, status=400)
            self.assertIn('Required', r.json['message'])

    def test_fails_if_state_does_not_match(self):
        self.app.app.registry.cache.set('def', 'http://foobar', ttl=1)
        url = '{url}?state=abc&code=1234'.format(url=self.url)
        resp = self.app.get(url, status=408)
        error_msg = 'The Facebook Auth session was not found, please re-authenticate.'
        self.assertFormattedError(
            resp, 408, ERRORS.MISSING_AUTH_TOKEN, "Request Timeout", error_msg)

    def test_fails_if_state_was_already_consumed(self):
        self.app.app.registry.cache.set('abc', 'http://foobar', ttl=1)
        url = '{url}?state=abc&code=1234'.format(url=self.url)
        self.app.get(url)
        resp = self.app.get(url, status=408)
        error_msg = 'The Facebook Auth session was not found, please re-authenticate.'
        self.assertFormattedError(
            resp, 408, ERRORS.MISSING_AUTH_TOKEN, "Request Timeout", error_msg)

    def test_fails_if_state_has_expired(self):
        with mock.patch.dict(self.app.app.registry.settings,
                             [('facebook.cache_ttl_seconds', 0.01)]):
            r = self.app.get(self.login_url)
        url = r.headers['Location']
        url_fragments = urlparse(url)
        queryparams = parse_qs(url_fragments.query)
        state = queryparams['state'][0]
        url = '{url}?state={state}&code=1234'.format(state=state, url=self.url)
        sleep(0.02)
        resp = self.app.get(url, status=408)
        error_msg = 'The Facebook Auth session was not found, please re-authenticate.'
        self.assertFormattedError(
            resp, 408, ERRORS.MISSING_AUTH_TOKEN, "Request Timeout", error_msg)

    def tests_redirects_with_token_traded_against_code(self):
        self.app.app.registry.cache.set('abc', 'http://foobar?token=', ttl=1)
        url = '{url}?state=abc&code=1234'.format(url=self.url)
        r = self.app.get(url)
        self.assertEqual(r.status_code, 302)
        self.assertEqual(r.headers['Location'],
                         'http://foobar?token=oauth-token')

    def tests_return_503_if_facebook_auth_server_behaves_badly(self):
        self.facebook_trade.get.return_value.raise_for_status.side_effect = HTTPError
        self.facebook_trade.exceptions.HTTPError = HTTPError

        self.app.app.registry.cache.set('abc', 'http://foobar', ttl=1)
        url = '{url}?state=abc&code=1234'.format(url=self.url)
        self.app.get(url, status=503)

    def tests_return_400_if_client_error_detected(self):
        httpBadRequest = mock.MagicMock()
        httpBadRequest.status_code = 400
        httpBadRequest.json.return_value = {"error": "blah"}
        self.facebook_trade.get.return_value = httpBadRequest

        self.app.app.registry.cache.set('abc', 'http://foobar', ttl=1)
        url = '{url}?state=abc&code=1234'.format(url=self.url)
        self.app.get(url, status=400)


class CapabilityTestView(BaseWebTest, unittest.TestCase):

    def test_facebook_capability(self, additional_settings=None):
        resp = self.app.get('/')
        capabilities = resp.json['capabilities']
        self.assertIn('facebook', capabilities)
        expected = {
            "version": kinto_facebook_version,
            "url": "https://github.com/Kinto/kinto-facebook",
            "description": "You can authenticate to that server using "
                           "Facebook Auth."
        }
        self.assertEqual(expected, capabilities['facebook'])


tests/test_includeme.py
import unittest

import kinto.core
from pyramid.exceptions import ConfigurationError
from pyramid import testing

from kinto_facebook import includeme


class IncludeMeTest(unittest.TestCase):
    def test_include_fails_if_kinto_was_not_initialized(self):
        config = testing.setUp()
        with self.assertRaises(ConfigurationError):
            config.include(includeme)

    def test_settings_are_filled_with_defaults(self):
        config = testing.setUp()
        kinto.core.initialize(config, '1.0.0')
        config.include(includeme)
        settings = config.get_settings()
        self.assertIsNotNone(settings.get('facebook.relier.enabled'))

    def test_a_heartbeat_is_registered_at_oauth(self):
        config = testing.setUp()
        kinto.core.initialize(config, '1.0.0')
        config.registry.heartbeats = {}
        config.include(includeme)
        self.assertIsNotNone(config.registry.heartbeats.get('facebook'))


tests/test_authentication.py
import time
import unittest

import mock
import pytest
import requests
from pyramid import httpexceptions

from kinto.core.cache import memory as memory_backend
from kinto.core.testing import DummyRequest

from kinto_facebook import authentication, DEFAULT_SETTINGS


class FacebookAuthenticationPolicyTest(unittest.TestCase):
    def setUp(self):
        self.policy = authentication.FacebookAuthenticationPolicy()
        self.backend = memory_backend.Cache(cache_prefix="tests",
                                            cache_max_size_bytes=float("inf"))

        self.request = self._build_request()
        self.profile_data = {
            "app_id": 138483919580948,
            "type": "USER",
            "application": "Social Cafe",
            "expires_at": 1352419328,
            "is_valid": True,
            "issued_at": 1347235328,
            "metadata": {
                "sso": "iphone-safari"
            },
            "scopes": [
                "email",
                "publish_actions"
            ],
            "user_id": "33"
        }

    def tearDown(self):
        self.backend.flush()

    def _build_request(self):
        request = DummyRequest()
        request.bound_data = {}
        request.registry.cache = self.backend
        settings = DEFAULT_SETTINGS.copy()
        settings['userid_hmac_secret'] = '1234'
        settings['facebook.cache_ttl_seconds'] = '0.01'
        request.registry.settings = settings
        request.headers['Authorization'] = 'Bearer foo'
        return request

    def test_returns_none_if_authorization_header_is_missing(self):
        self.request.headers.pop('Authorization')
        user_id = self.policy.unauthenticated_userid(self.request)
        self.assertIsNone(user_id)

    def test_returns_none_if_token_is_malformed(self):
        self.request.headers['Authorization'] = 'Bearerfoo'
        user_id = self.policy.unauthenticated_userid(self.request)
        self.assertIsNone(user_id)

    def test_returns_none_if_token_is_unknown(self):
        self.request.headers['Authorization'] = 'Carrier foo'
        user_id = self.policy.authenticated_userid(self.request)
        self.assertIsNone(user_id)

    @mock.patch('kinto_facebook.authentication.requests')
    def test_returns_facebook_userid(self, api_mocked):
        api_mocked.get.return_value.json.return_value = {"data": self.profile_data}
        user_id = self.policy.authenticated_userid(self.request)
        assert "33" == user_id

    @mock.patch('kinto_facebook.authentication.requests')
    def test_returns_facebook_user_id_in_principals(self, api_mocked):
        api_mocked.get.return_value.json.return_value = {"data": self.profile_data}
        principals = self.policy.effective_principals(self.request)
        self.assertIn("33", principals)

    @mock.patch('kinto_facebook.authentication.requests')
    def test_oauth_handle_invalid_tokens(self, api_mocked):
        api_mocked.get.return_value.json.return_value = {"data": {'is_valid': False}}
        principals = self.policy.effective_principals(self.request)
        self.assertNotIn("33", principals)

    @mock.patch('kinto_facebook.authentication.requests')
    def test_oauth_verification_is_cached(self, api_mocked):
        api_mocked.get.return_value.json.return_value = {"data": self.profile_data}
        # First request from client.
        request = self._build_request()
        self.policy.authenticated_userid(request)
        # Second request from same client.
        request = self._build_request()
        self.policy.authenticated_userid(request)
        # Cache backend was used.
        self.assertEqual(1, api_mocked.get.call_count)

    @mock.patch('kinto_facebook.authentication.requests')
    def test_oauth_verification_is_done_once_per_request(self, api_mocked):
        api_mocked.get.return_value.json.return_value = {"data": self.profile_data}
        # First request from client.
        self.policy.authenticated_userid(self.request)
        # Within the same request cycle, token won't be verified.
        self.request.headers['Authorization'] = 'Bearer another'
        self.policy.authenticated_userid(self.request)
        # Request bound data is used.
        self.assertEqual(1, api_mocked.get.call_count)

    @mock.patch('kinto_facebook.authentication.requests')
    def test_oauth_verification_uses_cache_by_token(self, api_mocked):
        api_mocked.get.return_value.json.return_value = {"data": self.profile_data}
        # First request from client.
        request = self._build_request()
        self.policy.authenticated_userid(request)
        # Second request from another client.
        request = self._build_request()
        request.headers['Authorization'] = 'Bearer another'
        self.policy.authenticated_userid(request)
        # Cache backend key was different.
        self.assertEqual(2, api_mocked.get.call_count)

    @mock.patch('kinto_facebook.authentication.requests')
    def test_oauth_verification_cache_has_ttl(self, api_mocked):
        api_mocked.get.return_value.json.return_value = {"data": self.profile_data}
        # First request from client.
        request = self._build_request()
        self.policy.authenticated_userid(request)
        # Second request from same client after TTL.
        time.sleep(0.02)
        request = self._build_request()
        self.policy.authenticated_userid(request)
        # Cache backend key was expired.
        self.assertEqual(2, api_mocked.get.call_count)

    @mock.patch('kinto_facebook.authentication.requests')
    def test_handle_facebook_server_misbehaviours(self, request_mock):
        request_mock.get.return_value.raise_for_status.side_effect = requests.exceptions.HTTPError
        # Since we mock requests, we need to provide requests.exceptions here too
        request_mock.exceptions.HTTPError = requests.exceptions.HTTPError

        with pytest.raises(httpexceptions.HTTPServiceUnavailable):
            request = self._build_request()
            self.policy.authenticated_userid(request)

    def test_forget_uses_realm(self):
        policy = authentication.FacebookAuthenticationPolicy(realm='Who')
        headers = policy.forget(self.request)
        self.assertEqual(headers[0],
                         ('WWW-Authenticate', 'Bearer realm="Who"'))


class FacebookPingTest(unittest.TestCase):
    def setUp(self):
        self.request = DummyRequest()
        self.request.registry.settings = DEFAULT_SETTINGS
        self.request.registry.settings['facebook.authorization_endpoint'] = 'http://graph.facebook'

    def test_returns_none_if_oauth_deactivated(self):
        self.request.registry.settings['facebook.authorization_endpoint'] = None
        self.assertIsNone(authentication.facebook_ping(self.request))

    @mock.patch('requests.get')
    def test_returns_true_if_ok(self, get_mocked):
        httpOK = requests.models.Response()
        httpOK.status_code = 200
        get_mocked.return_value = httpOK
        self.assertTrue(authentication.facebook_ping(self.request))

    @mock.patch('requests.get')
    def test_returns_false_if_ko(self, get_mocked):
        get_mocked.side_effect = requests.exceptions.HTTPError()
        self.assertFalse(authentication.facebook_ping(self.request))


kinto_facebook/__init__.py
import pkg_resources

from pyramid.exceptions import ConfigurationError
from pyramid.settings import asbool

from kinto_facebook.authentication import facebook_ping

#: Module version, as defined in PEP-0396.
__version__ = pkg_resources.get_distribution(__package__).version


DEFAULT_SETTINGS = {
    'facebook.authorization_endpoint': 'https://www.facebook.com/v2.12/dialog/oauth',
    'facebook.token_endpoint': 'https://graph.facebook.com/v2.12/oauth/access_token',
    'facebook.userinfo_endpoint': 'https://graph.facebook.com/debug_token',
    'facebook.app_access_token': None,
    'facebook.client_id': None,
    'facebook.client_secret': None,
    'facebook.relier.enabled': True,
    'facebook.header_type': 'Bearer',
    'facebook.requested_scope': 'public_profile email',
    'facebook.cache_ttl_seconds': 5 * 60,
    'facebook.state.ttl_seconds': 3600,  # 1 hour
    'facebook.heartbeat_timeout_seconds': 3,
    'facebook.webapp.authorized_domains': '',
}


def includeme(config):
    if not hasattr(config.registry, 'heartbeats'):
        message = ('kinto-facebook should be included once Kinto is initialized. '
                   'Use setting ``kinto.includes`` instead of ``pyramid.includes`` '
                   'or include it manually.')
        raise ConfigurationError(message)

    settings = config.get_settings()

    defaults = {k: v for k, v in DEFAULT_SETTINGS.items() if k not in settings}
    config.add_settings(defaults)

    # Register heartbeat to ping Facebook server.
    config.registry.heartbeats['facebook'] = facebook_ping

    config.add_api_capability(
        "facebook",
        version=__version__,
        description="You can authenticate to that server "
                    "using Facebook Auth.",
        url="https://github.com/Kinto/kinto-facebook")

    # Ignore Facebook Auth relier endpoint in case it's not activated.
    relier_enabled = asbool(settings['facebook.relier.enabled'])
    if relier_enabled:
        config.scan('kinto_facebook.views')


kinto_facebook/authentication.py
import logging

import requests
from pyramid import authentication as base_auth
from pyramid import httpexceptions
from pyramid.interfaces import IAuthenticationPolicy
from zope.interface import implementer

from kinto.core import utils as core_utils
from kinto_facebook.utils import facebook_conf

logger = logging.getLogger(__name__)

REIFY_KEY = 'facebook_verified_token'


@implementer(IAuthenticationPolicy)
class FacebookAuthenticationPolicy(base_auth.CallbackAuthenticationPolicy):
    def __init__(self, realm='Realm'):
        self.realm = realm
        self._cache = None

    def unauthenticated_userid(self, request):
        """Return the Facebook user_id or ``None`` if token could not be verified.
        """
        authorization = request.headers.get('Authorization', '')
        try:
            authmeth, token = authorization.split(' ', 1)
        except ValueError:
            return None
        if authmeth.lower() != facebook_conf(request, 'header_type').lower():
            return None

        user_id = self._verify_token(token, request)
        return user_id

    def forget(self, request):
        """A no-op. Credentials are sent on every request.
        Return WWW-Authenticate Realm header for Bearer token.
        """
        return [('WWW-Authenticate', '{} realm="{}"'.format(
            facebook_conf(request, 'header_type'), self.realm))]

    def _verify_token(self, access_token, request):
        """Verify the token extracted from the Authorization header.

        This method stores the result in two locations to avoid hitting the
        auth remote server as much as possible:

        - on the request object, in case the Pyramid authentication methods
          like `effective_principals()` or `authenticated_userid()` are called
          several times during the request cycle;

        - in the cache backend, to reuse validated token from one request to
          another (during ``cache_ttl_seconds`` seconds.)
        """
        # First check if this request was already verified.
        # `request.bound_data` is an attribute provided by Kinto to store
        # some data that is shared among sub-requests (e.g. default bucket
        # or batch requests)
        if REIFY_KEY not in request.bound_data:
            settings = request.registry.settings
            hmac_secret = settings['userid_hmac_secret']

            cache_ttl = float(facebook_conf(request, 'cache_ttl_seconds'))

            hmac_token = core_utils.hmac_digest(hmac_secret, access_token)
            cache_key = 'facebook:verify:{}'.format(hmac_token)

            payload = request.registry.cache.get(cache_key)

            if payload is None:
                # Verify token from Facebook
                url = facebook_conf(request, 'userinfo_endpoint')
                params = {
                    'input_token': access_token,
                    'access_token': facebook_conf(request, 'app_access_token'),
                }

                # XXX: Implement token validation for Facebook
                resp = requests.get(url, params=params)
                try:
                    resp.raise_for_status()
                except requests.exceptions.HTTPError as e:
                    logger.exception("Facebook Token Protocol Error")
                    raise httpexceptions.HTTPServiceUnavailable()
                else:
                    body = resp.json()
                    if not body['data']['is_valid']:
                        payload = {}
                    else:
                        payload = body['data']

            request.registry.cache.set(cache_key, payload, ttl=cache_ttl)

            # Save for next call.
            request.bound_data[REIFY_KEY] = payload.get('user_id')

        return request.bound_data[REIFY_KEY]


def facebook_ping(request):
    """Verify if the OAuth server is ready."""
    heartbeat_url = facebook_conf(request, 'authorization_endpoint')

    facebook = None
    if heartbeat_url is not None:
        try:
            timeout = float(facebook_conf(request, 'heartbeat_timeout_seconds'))
            r = requests.get(heartbeat_url, timeout=timeout)
            r.raise_for_status()
            facebook = True
        except requests.exceptions.HTTPError:
            pass

    return facebook


kinto_facebook/views.py
import logging
import uuid
from fnmatch import fnmatch
from urllib.parse import urlparse, urlencode

import colander
import requests
from cornice.validators import colander_validator
from pyramid import httpexceptions
from pyramid.security import NO_PERMISSION_REQUIRED
from pyramid.settings import aslist

from kinto.core import Service
from kinto.core.errors import http_error, ERRORS, json_error_handler, raise_invalid
from kinto.core.resource.schema import URL

from kinto_facebook.utils import facebook_conf


logger = logging.getLogger(__name__)


login = Service(name='facebook-login',
                path='/facebook/login',
                error_handler=json_error_handler)

token = Service(name='facebook-token',
                path='/facebook/token',
                error_handler=json_error_handler)


def persist_state(request):
    """Persist arbitrary string in cache.
    It will be matched when the user returns from the OAuth server login
    page.
    """
    state = uuid.uuid4().hex
    redirect_url = request.validated['querystring']['redirect']
    expiration = float(facebook_conf(request, 'cache_ttl_seconds'))

    request.registry.cache.set(state, redirect_url, expiration)

    return state


class FacebookLoginQueryString(colander.MappingSchema):
    redirect = URL()


class FacebookLoginRequest(colander.MappingSchema):
    querystring = FacebookLoginQueryString()


def authorized_redirect(req, **kwargs):
    authorized = aslist(facebook_conf(req, 'webapp.authorized_domains'))
    if not req.validated:
        # Schema was not validated. Give up.
        return False

    redirect = req.validated['querystring']['redirect']

    domain = urlparse(redirect).netloc

    if not any((fnmatch(domain, auth) for auth in authorized)):
        req.errors.add('querystring', 'redirect',
                       'redirect URL is not authorized')


@login.get(schema=FacebookLoginRequest, permission=NO_PERMISSION_REQUIRED,
           validators=(colander_validator, authorized_redirect))
def facebook_login(request):
    """Helper to redirect client towards Facebook login form."""
    state = persist_state(request)

    params = {
        'client_id': facebook_conf(request, 'client_id'),
        'redirect_uri': request.route_url(token.name),
        'state': state
    }

    login_form_url = '{}?{}'.format(facebook_conf(request, 'authorization_endpoint'),
                                    urlencode(params))

    request.response.status_code = 302
    request.response.headers['Location'] = login_form_url

    return {}


class OAuthQueryString(colander.MappingSchema):
    code = colander.SchemaNode(colander.String())
    state = colander.SchemaNode(colander.String())


class OAuthRequest(colander.MappingSchema):
    querystring = OAuthQueryString()


@token.get(schema=OAuthRequest, permission=NO_PERMISSION_REQUIRED,
           validators=(colander_validator,))
def facebook_token(request):
    """Return OAuth token from authorization code.
    """
    state = request.validated['querystring']['state']
    code = request.validated['querystring']['code']

    # Require on-going session
    stored_redirect = request.registry.cache.get(state)

    # Make sure we cannot try twice with the same code
    request.registry.cache.delete(state)
    if not stored_redirect:
        error_msg = 'The Facebook Auth session was not found, please re-authenticate.'
        return http_error(httpexceptions.HTTPRequestTimeout(),
                          errno=ERRORS.MISSING_AUTH_TOKEN,
                          message=error_msg)

    url = facebook_conf(request, 'token_endpoint')
    params = {
        'client_id': facebook_conf(request, 'client_id'),
        'client_secret': facebook_conf(request, 'client_secret'),
        'redirect_uri': request.route_url(token.name),
        'code': code,
    }

    resp = requests.get(url, params=params)
    if resp.status_code == 400:
        response_body = resp.json()
        logger.error("Facebook Token Validation Failed: {}".format(response_body))
        error_details = {
            'name': 'code',
            'location': 'querystring',
            'description': 'Facebook OAuth code validation failed.'
        }
        raise_invalid(request, **error_details)

    try:
        resp.raise_for_status()
    except requests.exceptions.HTTPError as e:
        logger.exception("Facebook Token Protocol Error")
        raise httpexceptions.HTTPServiceUnavailable()
    else:
        response_body = resp.json()
        access_token = response_body['access_token']

    return httpexceptions.HTTPFound(location='%s%s' % (stored_redirect, access_token))
